File: 0098-Validate-Binary-Search-Tree/java-0098/src/Solution3.java
/// Source : https://leetcode.com/problems/validate-binary-search-tree/description/
/// Author : liuyubobobo
/// Time   : 2018-12-16


import java.util.Stack;

/// Non-Recursion Solution for Recursion thinking
/// Time Complexity: O(n)
/// Space Complexity: O(h)
public class Solution3 {


    public boolean isValidBST(TreeNode root) {

        if(root == null) return true;

        Stack<TreeNode> stack = new Stack<>();
        Stack<Integer> lower = new Stack<>();
        Stack<Integer> upper = new Stack<>();

        stack.push(root);
        lower.push(Integer.MIN_VALUE);
        upper.push(Integer.MAX_VALUE);
        while(!stack.empty()){
            TreeNode cur = stack.pop();
            int left = lower.pop();
            int right = upper.pop();

            if(cur.val < left || cur.val > right)
                return false;

            if(cur.right != null){
                if(cur.right.val <= cur.val) return false;
                stack.push(cur.right);
                lower.push(cur.val + 1);
                upper.push(right);
            }

            if(cur.left != null){
                if(cur.left.val >= cur.val) return false;
                stack.push(cur.left);
                lower.push(left);
                upper.push(cur.val - 1);
            }
        }
        return true;
    }
}

File: 0896-Monotonic-Array/py-0896/Solution3.py
# Source : https://leetcode.com/problems/monotonic-array/
# Author : penpenps
# Time   : 2019-07-29

from typing import List

# One-line solution
# Time Complexity: O(n)
# Space Complexity: O(1)

class Solution:
    def isMonotonic(self, A: List[int]) -> bool:
        return not {(x>y) - (y>x) for x, y in zip(A, A[1:])} >= {1, -1}